arci-ros2/tests/test_navigation.rs

```diff
@@ -0,0 +1,197 @@
+        .unwrap();
+
+    let node_cb = node.clone();
+    tokio::spawn(test_nav_server(node_cb, server_requests));
+
+    std::thread::spawn(move || loop {
+        node.lock().unwrap().spin_once(Duration::from_millis(100));
+    });
+
+    nav.send_goal_pose(
+        Isometry2::new(Vector2::new(-0.6, 0.2), 1.0),
+        "map",
+        Duration::from_secs(10),
+    )
+    .unwrap()
+    .await
+    .unwrap();
+}
+
+#[tokio::test]
+async fn test_nav_cancel() {
+    const ACTION_NAME: &str = "/test_nav_cancel";
+
+    let ctx = r2r::Context::create().unwrap();
+    let nav = Ros2Navigation::new(ctx.clone(), ACTION_NAME);
+    let node = Arc::new(Mutex::new(
+        r2r::Node::create(ctx, "nav2_node", "arci_ros2").unwrap(),
+    ));
+
+    let server_requests = node
+        .lock()
+        .unwrap()
+        .create_action_server::<NavigateToPose::Action>(ACTION_NAME)
+        .unwrap();
+
+    let node_cb = node.clone();
+    tokio::spawn(test_nav_server(node_cb, server_requests));
+
+    std::thread::spawn(move || loop {
+        node.lock().unwrap().spin_once(Duration::from_millis(100));
+    });
+
+    let wait = nav
+        .send_goal_pose(
+            Isometry2::new(Vector2::new(-0.6, 0.2), 1.0),
+            "map",
+            Duration::from_secs(10),
+        )
+        .unwrap();
+    // TODO: remove needs of this sleep
+    tokio::time::sleep(Duration::from_millis(1000)).await;
+    nav.cancel().unwrap();
+    // TODO: "canceled" should be an error?
+    wait.await.unwrap();
+}
+
+async fn run_goal(
+    node: Arc<Mutex<r2r::Node>>,
+    g: r2r::ActionServerGoal<NavigateToPose::Action>,
+    canceled: Arc<AtomicBool>,
+) -> NavigateToPose::Result {
+    let mut timer = node // local timer, will be dropped after this request is processed.
+        .lock()
+        .unwrap()
+        .create_wall_timer(Duration::from_millis(1000))
+        .expect("could not create timer");
+
+    let mut feedback_msg = NavigateToPose::Feedback::default();
+    g.publish_feedback(feedback_msg.clone()).expect("fail");
+
+    let mut orientation_diff = feedback_msg.current_pose.pose.orientation.clone();
+    orientation_diff.w = (g.goal.pose.pose.orientation.w - orientation_diff.w) / 10.0;
+    orientation_diff.x = (g.goal.pose.pose.orientation.x - orientation_diff.x) / 10.0;
+    orientation_diff.y = (g.goal.pose.pose.orientation.y - orientation_diff.y) / 10.0;
+    orientation_diff.z = (g.goal.pose.pose.orientation.z - orientation_diff.z) / 10.0;
+    let mut position_diff = feedback_msg.current_pose.pose.position.clone();
+    position_diff.x = (g.goal.pose.pose.position.x - position_diff.x) / 10.0;
+    position_diff.y = (g.goal.pose.pose.position.y - position_diff.y) / 10.0;
+    position_diff.z = (g.goal.pose.pose.position.z - position_diff.z) / 10.0;
+    for _ in 0..10 {
+        if canceled.load(Ordering::SeqCst) {
+            break;
+        }
+        feedback_msg.current_pose.pose.orientation.w += orientation_diff.w;
```

Comment:
I think it's better to use some variables to make the code shorter.

```rust
let fb_pose = &feedback_msg.current_pose.pose;
```

especially for L129-158

Reply:
done